apis/Google.Longrunning/Google.Longrunning/Operation.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Protobuf;
+using Grpc.Core;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Longrunning
+{
+    /// <summary>
+    /// A long-running operation with an associated client, and which knows the expected response type.
+    /// </summary>
+    /// <remarks>
+    /// For simplicity, no methods on this type modify the proto message. Instead, to get up-to-date
+    /// information you can use Refresh to obtain a new instance.
+    /// </remarks>
+    /// <typeparam name="T">The response message type.</typeparam>
+    public class Operation<T> where T : IMessage<T>, new()
+    {
+        /// <summary>
+        /// The protobuf message associated with the long-running operation, containing the name (for
+        /// further retrieval) and any error/result already computed. This should not be mutated.
+        /// </summary>
+        public Operation Proto { get; }
+        public OperationsClient Client { get; }
+
+        public Operation(Operation proto, OperationsClient client)
+        {
+            Proto = GaxPreconditions.CheckNotNull(proto, nameof(proto));
+            Client = GaxPreconditions.CheckNotNull(client, nameof(proto));
+        }
+
+        /// <summary>
+        /// Returns the name of the operation, which can be persisted and used to poll for the latest
+        /// results at a later time or in a different program.
+        /// </summary>
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// </remarks>
+        public string Name => Proto.Name;
+
+        /// <summary>
+        /// Whether the operation has completed, where "complete" can include "failed".
+        /// </summary>
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// </remarks>
+        public bool IsCompleted => Proto.Done;
+
+        /// <summary>
+        /// Whether the operation has completed with a failure.
+        /// </summary>
+        public bool IsFaulted => Proto.ResultCase == Operation.ResultOneofCase.Error;
+
+        // TODO: An Exception property? Would probably want to use a Lazy<OperationFailedException> to avoid creating
```

Comment:
Yes, I'd think that an Exception property would be useful. Possibly lazy as you suggest.